morebits.js

```diff
@@ -2879,36 +2923,121 @@ Morebits.wiki.page = function(pageName, currentAction) {
 
 		// check for "Database query error"
 		if ( errorCode === "internal_api_error_DBQueryError" && ctx.retries++ < ctx.maxRetries ) {
-
 			ctx.statusElement.info("Database query error, retrying");
 			--Morebits.wiki.numberOfActionsLeft;  // allow for normal completion if retry succeeds
 			ctx.deleteProcessApi.post(); // give it another go!
-
 		} else if ( errorCode === "badtoken" ) {
 			// this is pathetic, but given the current state of Morebits.wiki.page it would
 			// be a dog's breakfast to try and fix this
 			ctx.statusElement.error("Invalid token. Please refresh the page and try again.");
 			if (ctx.onDeleteFailure) {
 				ctx.onDeleteFailure.call(this, this, ctx.deleteProcessApi);
 			}
-
 		} else if ( errorCode === "missingtitle" ) {
-
 			ctx.statusElement.error("Cannot delete the page, because it no longer exists");
 			if (ctx.onDeleteFailure) {
 				ctx.onDeleteFailure.call(this, ctx.deleteProcessApi);  // invoke callback
 			}
-
 		// hard error, give up
 		} else {
-
 			ctx.statusElement.error( "Failed to delete the page: " + ctx.deleteProcessApi.getErrorText() );
 			if (ctx.onDeleteFailure) {
 				ctx.onDeleteFailure.call(this, ctx.deleteProcessApi);  // invoke callback
 			}
 		}
 	};
 
+	var fnProcessUndelete = function() {
+		var pageTitle, token;
+
+		// The whole handling of tokens in Morebits is outdated (#615)
+		// but has generally worked since intoken has been deprecated
+		// but remains.  intoken does not, however, take undelete.
+		// Everything except watching and patrolling should eventually
+		// use csrf, but until then (#615) the stupid hack below
+		// should work for undeletion
+		if (fnCanUseMwUserToken('undelete')) {
```

Comment:
I don't understand, `token = mw.user.tokens.get('editToken');` is being used whether or not this condition is true. Why have this condition at all?

Reply:
This is what I was trying to convey in the comment: until I can radically modernize the handling of tokens, I wanted to keep the code structure parallel.  I think it's easier to understand and fix/remove/rewrite if as much as possible follows a similar, understandable pathway rather than a bespoke system for each.  Yes, we can remove this `fnCanUseMwUserToken`, I suppose that'd be fine given the context of the comment.

Reply:
Thought I [dealt](https://github.com/azatoth/twinkle/compare/f4b189e7d85639f50c6d9f40ef5c2b51d06b3319..4f85861eda659e8f8bb1be510b61003552189ee7) with this, but that meant morebits failed, so I've [undone](https://github.com/azatoth/twinkle/compare/2a5e111dc294abc635f4ebad5d9f6a497ce69a2b..a738d5b30e32ba1115013b86f5158652dd24c205) it for now as I don't have any more time today.

Reply:
Oh duh!  If `suppressProtectWarning` isn't set, the pages fail the "poor man's normalisation" in `fnCanUseMwUserToken`.  `suppressProtectWarning` is set for the batch but not for talkpages so the kludge was only reached (in this case) for talk pages.  This matches the behavior in `twinklebatchdelete`, and is required; since the batches check, display, and confirm on protection status for the pages listed (well, *now* `batchundelete` does) it's not needed for those, but the confirms(s) are needed for any talkpages found later.  Sorry for leading you on @siddharthvp, I knew this a few days ago but `Morebits` gets confusing!

As far as `fnCanUseMwUserToken` goes, `followRedirect` doesn't really make sense for `undeletePage` so the `suppressProtectWarning` is the real killer.  The real goal of that whole thing seems to be to avoid another API call to get the protection status if `suppressProtectWarning` isn't set, none of that is relevant for batch-processing like this.  I'm inclined to leave it as is for now on the off chance we decide to use it elsewhere before revamping `morebits`, as well as, as mentioned elsewhere, making what will surely be an revamping process easier.